File: hw2/src/Moon.cpp
#include "Moon.h"
#include <stdexcept>

namespace moon {

    Moon::Moon(int size, std::string color, int phase) {
        m_size = size;
        m_color = color;
        m_phase = phase;
    }
    bool Moon::isNewMoon() {
        return m_phase == 0;
    }
    void Moon::cycle() {
        m_phase = (m_phase + 1) % 4;
    }
    int Moon::phase() {
        return m_phase;
    }
    std::string Moon::color() {
        return m_color;
    }
    void Moon::color(std::string value) {
        if (!(value == "silver" || value == "white" )) {
            throw std::invalid_argument("invalide color");
        }
        m_color = value;
    }
    
    int Moon::size() {
        return m_size;
    }

}

File: js-oo/moon.mjs
export class Moon {
    constructor(size, color, phase) {
        this._size = size
        this._color = color
        this._phase = phase
    }

    config(div,id,config) {
        div.innerHTML="Moon(id=" + id + ")";
    }

    isNewMoon() {
        return this._phase == 0
    }

    cycle() {
        this._phase = (this._phase + 1) % 4
    }

    get phase() {
        return this._phase
    }

    get color() {
        return this._color
    }

    set color(value) {
        if (! ["white","silver"].includes(value)) {
            throw Error('invalid value')
        }
        this._color = value
    }

    get size() {
        return this._size
    }
}

function MoonInstantiate() {
    const divs=document.getElementsByClassName("moon");
    for (let i=0; i<divs.length; ++i) {
        const div=divs[i];
        const id = "moon-" + i;
        const config = JSON.parse(div.getAttribute("data-config"));
        const moon = new Moon(div,id,config)
    }
}
